File: GripperX/Software/ros2/src/gripperx_swerve_controller/test/dump_reference.py
#!/usr/bin/env python3
r"""
Functional-equivalence harness, PYTHON side (NFR-10 acceptance 1).

Reads "vx vy omega d0 d1 d2 d3" per line from stdin (current steering angles in
MODEL order FL, BL, BR, FR, radians) and prints what today's chain resolves for
that twist, using the LIVE modules the running robot uses — no reimplementation:

    gripperx_control.swerve_kinematic_model.FourWIS4WIDKinematicModel
    gripperx_control.steering_limits.limit_twist_to_steering_range
    swerve_cmd_node._steer_alignment_scale  (inlined below: importing
        swerve_cmd_node pulls in rclpy, which this harness does not need)

Output per case, matching test/dump_reference.cpp:

    <status> a0 s0 c0 a1 s1 c1 a2 s2 c2 a3 s3 c3

with a = commanded steering angle, s = commanded wheel angular velocity,
c = the steer-alignment (slew braking) scale that produced s.

There is deliberately NO --reference switch here, unlike on the C++ side.
swerve_cmd_node cannot see arbitration point A2 — the /teleop/direct_steer
override lives in another process — so the IK target is the only braking
reference today's chain HAS. That asymmetry is the point of the comparison:
`dump_reference --reference=commanded` must reproduce this output exactly on
every case without override columns, and is expected to differ on cases with
them. Any line with more than 7 numbers therefore has its extra columns ignored
here rather than silently changing this side's behaviour.

Run it against test/dump_reference.cpp's output with the same input; any
difference beyond floating-point noise is a failed port.

Usage:
    PYTHONPATH=<repo>/Software/ros2/src/gripperx_control/src \\
        python3 dump_reference.py < cases.txt
"""

import math
import sys

from gripperx_control.steering_limits import (
    DEFAULT_INWARD_LIMIT_DEG,
    DEFAULT_OUTWARD_LIMIT_DEG,
    DEFAULT_OUTWARD_SIGN,
    limit_twist_to_steering_range,
    LimitStatus,
    normalize_angle,
    SteeringLimits,
)
from gripperx_control.swerve_kinematic_model import (
    BodyTwist,
    FourWIS4WIDKinematicModel,
)

# gripperx_control/config/swerve_cmd.yaml, unchanged.
from gripperx_geometry.constants import (  # noqa: E402
    HALF_TRACK_KINGPIN as B,
    HALF_WHEELBASE_KINGPIN as A,
    WHEEL_RADIUS_EFFECTIVE as R,
)


MIN_SCALE = 0.45
DEADBAND = 0.12
REFERENCE = 1.0472
MAX_WHEEL_ANGULAR_SPEED = 12.0


def alignment_scale(target: float, current: float) -> float:
    error = abs(normalize_angle(target - current))
    if error <= DEADBAND:
        return 1.0
    return max(MIN_SCALE, 1.0 - (error / REFERENCE))


def main() -> int:
    model = FourWIS4WIDKinematicModel(a=A, b=B, wheel_radius=R)
    limits = SteeringLimits.from_outward_inward(
        math.radians(DEFAULT_OUTWARD_LIMIT_DEG),
        math.radians(DEFAULT_INWARD_LIMIT_DEG),
        DEFAULT_OUTWARD_SIGN,
    ).in_model_order()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        values = [float(v) for v in line.split()]
        vx, vy, omega = values[0:3]
        current = values[3:7]

        limited = limit_twist_to_steering_range(
            model, BodyTwist(vx=vx, vy=vy, omega=omega), current, limits
        )
        if limited.status == LimitStatus.REJECTED:
            print('rejected')
            continue

        parts = [limited.status]
        for target, current_angle in zip(limited.targets, current):
            scale = alignment_scale(target.angle, current_angle)
            speed = (target.speed / R) * scale
            speed = max(-MAX_WHEEL_ANGULAR_SPEED, min(MAX_WHEEL_ANGULAR_SPEED, speed))
            parts.append('%.9f' % target.angle)
            parts.append('%.9f' % speed)
            parts.append('%.9f' % scale)
        print(' '.join(parts))
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: GripperX/Software/ros2/src/gripperx_swerve_controller/test/dump_reference.cpp
// Functional-equivalence harness, C++ side (NFR-10 acceptance 1).
//
// Reads one case per line from stdin:
//
//     vx vy omega d0 d1 d2 d3 [o0 o1 o2 o3]
//
// d0..d3 are the CURRENT steering angles in MODEL order (FL, BL, BR, FR),
// radians. o0..o3 are OPTIONAL and are the POST-ARBITRATION commanded steering
// angles for the same cycle, also in model order — i.e. what an active
// /teleop/direct_steer override (arbitration point A2) makes the controller
// command instead of the IK target. Their presence is what lets this harness
// exercise the reference change described below; without them a case is a
// plain no-override case.
//
// Prints, per case:
//
//     <status> a0 s0 c0 a1 s1 c1 a2 s2 c2 a3 s3 c3
//
// with a = commanded steering angle, s = commanded wheel angular velocity,
// c = the steer-alignment (slew braking) scale that produced s.
//
// --reference=ik         brake against the IK-derived target  (swerve_cmd_node's
//                        behaviour today; the DEFAULT, so an unqualified run of
//                        this harness still reproduces the old chain)
// --reference=commanded  brake against the post-arbitration commanded target
//                        (the controller's behaviour after the 2026-08-19 user
//                        decision)
//
// The two modes are IDENTICAL on every case that carries no override columns —
// there the commanded target IS the IK target. That is the property the 413-case
// comparison against test/dump_reference.py checks, and it is why the reference
// change does not invalidate the equivalence harness: it narrows its scope to
// the no-override cases and makes the override cases a separate, intended-
// difference set.
//
// test/dump_reference.py does the same through the LIVE Python modules
// (gripperx_control.steering_limits + swerve_kinematic_model + swerve_cmd_node's
// _steer_alignment_scale). It has no --reference switch on purpose: today's
// chain cannot see the override, so "ik" is the only reference it can offer.
// Diffing the two outputs is what makes "functionally equivalent to today's
// chain" a check rather than a claim.
//
// The steer-feedback differential is deliberately OUT of this harness: it is
// stateful (a first-order filter over dt), so it is not a pure function of one
// twist. It is ported line for line in the controller and reviewed there.

#include <cmath>
#include <cstdio>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "gripperx_swerve_controller/steering_limits.hpp"
#include "gripperx_swerve_controller/swerve_kinematics.hpp"

using namespace gripperx_swerve_controller;  // NOLINT(build/namespaces)

namespace
{
// Values from gripperx_control/config/swerve_cmd.yaml, unchanged.
constexpr double kA = 0.1809;
constexpr double kB = 0.1087;
constexpr double kR = 0.070;
constexpr double kMinScale = 0.45;
constexpr double kDeadband = 0.12;
constexpr double kReference = 1.0472;
constexpr double kMaxWheelAngularSpeed = 12.0;

double alignment_scale(double target, double current)
{
  const double error = std::fabs(normalize_angle(target - current));
  if (error <= kDeadband) {
    return 1.0;
  }
  return std::max(kMinScale, 1.0 - (error / kReference));
}
}  // namespace

int main(int argc, char ** argv)
{
  bool reference_is_commanded = false;
  for (int i = 1; i < argc; ++i) {
    if (std::strcmp(argv[i], "--reference=commanded") == 0) {
      reference_is_commanded = true;
    } else if (std::strcmp(argv[i], "--reference=ik") == 0) {
      reference_is_commanded = false;
    } else {
      std::fprintf(stderr, "unknown argument: %s\n", argv[i]);
      return 2;
    }
  }

  const SwerveKinematics model(kA, kB, kR);
  const auto limits = SteeringLimits::from_outward_inward(
                        kDefaultOutwardLimitDeg * M_PI / 180.0,
                        kDefaultInwardLimitDeg * M_PI / 180.0, kDefaultOutwardSign)
                        .in_model_order();

  std::string line;
  while (std::getline(std::cin, line)) {
    std::istringstream stream(line);
    std::vector<double> values;
    double value = 0.0;
    while (stream >> value) {
      values.push_back(value);
    }
    if (values.empty()) {
      continue;
    }
    if (values.size() != 7 && values.size() != 11) {
      std::fprintf(stderr, "expected 7 or 11 numbers per line, got %zu\n", values.size());
      return 2;
    }

    const BodyTwist twist{values[0], values[1], values[2]};
    std::array<double, kNumWheels> current{};
    for (std::size_t i = 0; i < kNumWheels; ++i) {
      current[i] = values[3 + i];
    }
    const bool override_active = values.size() == 11;
    std::array<double, kNumWheels> override_angles{};
    for (std::size_t i = 0; i < kNumWheels && override_active; ++i) {
      override_angles[i] = values[7 + i];
    }

    const auto limited = limit_twist_to_steering_range(model, twist, current, limits);
    const char * status = limited.status == LimitStatus::kOk             ? "ok"
                          : limited.status == LimitStatus::kOmegaReduced ? "omega_reduced"
                                                                        : "rejected";
    std::printf("%s", status);
    if (limited.status == LimitStatus::kRejected) {
      // Zero drive and the steering holds — no scale is computed, in the
      // controller or here, so the reference mode cannot change this line.
      std::printf("\n");
      continue;
    }
    for (std::size_t i = 0; i < kNumWheels; ++i) {
      const auto & target = (*limited.targets)[i];
      // The one line the whole reference change lives on. Without override
      // columns the two modes pick the same number.
      const double reference_angle = (reference_is_commanded && override_active)
                                       ? override_angles[i]
                                       : target.angle;
      const double scale = alignment_scale(reference_angle, current[i]);
      double speed = (target.speed / kR) * scale;
      speed = std::max(-kMaxWheelAngularSpeed, std::min(kMaxWheelAngularSpeed, speed));
      // The COMMANDED angle is the override when one is active; the IK target
      // otherwise. Printed so a diff shows both halves of a case, not just the
      // speed the scale produced.
      const double commanded_angle = override_active ? override_angles[i] : target.angle;
      std::printf(" %.9f %.9f %.9f", commanded_angle, speed, scale);
    }
    std::printf("\n");
  }
  return 0;
}
